### DISCUSSION TITLE
Importing/Mapping: Import works for withdrawals but fails for deposits

### DISCUSSION THREAD

**Discussion by sodaws** [firefly-iii/firefly-iii#6233] [Firefly III] [answered]
Hey,
I have a problem with fidi and mapping. Don't know if it's a bug or I'm using it wrong, therefore a discussion instead of an issue.

Checked the docs, disussions and issues but found nothing directly related to my problem. (#6165 could be related distantly)
I'm on the current stable release versions of both as of today. This should be reproduceable on a fresh install with one Asset Account created.

### What I'm trying to do
I'm trying to import my transactions into ff3 via fidi from a csv I downloaded from my bank. Importing without mapping works as expected.
The issue occurs while importing with "Opposing account (name)" mapping.

This is what I'm trying to do more specificaly:
In my csv, the names of the opposing account can have multiple variations (PayPal is one example).
I don't want to have an Expense Account and Revenue Account for each of these variations. Instead, I use mapping to assign all transactions from these multiple accounts to a single Expense/Revenue account (Intended use case if I understood the [docs ](https://docs.firefly-iii.org/data-importer/usage/map/) correctly).

To be speficic:
All transactions from **OR** to: 
* PayPal (Europe) S.a r.l. et Cie, S.C.A.
* PAYPAL EUROPE SARL ET CIE SCA
* PayPal (Europe) S.a.r.l. et Cie., S.C.A.
* PayPal Europe

shall be linked accordingly to an Expense or Revenue Account that's just called "PayPal" (Name Field of the account in Expense/Revenue Account Overview, Destination/Source Account in Transaction Overview).

### How to reproduce

To achieve this, I did this on a fresh install:
1. Create an Asset Account (Name: ExampleAssetAccount)
2. Create an Expense Account (Name: PayPal, other fields empty, since I'm grouping multiple accounts with different SEPA/BIC/AC into one account)
3. Create a Revenue Account (Name: PayPal, other fields empty, since I'm grouping multiple accounts with different SEPA/BIC/AC into one account)
4. Import data with fidi.

Step 4 is where the problem is:
* In the fidi role menu, I assinged the role Opposing Account (name) to its correct csv column and ticked the mapping box.
* In the fidi mapping menu, it's not clear to me to which exact account I'm supposed to map the field to.
I can select between Revenue, Expense, Cash and Asset Accounts from the list.
In my case, I assume I want to select one of the accounts that's just called "PayPal" but the question is:
  * "PayPal" under Revenue Account section?
  * or "PayPal" under Expense Account section?

Remember, the 4 opposing accounts from above that I'm trying to map to "PayPal" are both Expense and Revenue accounts at the same time. Meaning that there are transactions from AND to these accounts.

### The error

Ok, what I assumed here is that it doesn't matter whether I select the "PayPal" Expense account or "PayPal" Revenue account and the expenses or revenues are linked to the right account automatically.
So I just selected the "PayPal" Expense Account for all 4.
Unfortunately, that doesn't seem work.

When trying to do this, rows with expenses are imported correctly but for rows that are deposits/revenues ff3/fidi throws this error:

    Line #1:
    
        (2)transactions.0.source_id: Could not find a valid source account when searching for ID "5" or name "". (original value: "5")
        (2)transactions.0.source_name: Could not find a valid source account when searching for ID "5" or name "". (original value: "(not found)")

ID 5 is the "PayPal" Expense Account but it should be linked to ID 10, the "PayPal" Revenue Account

### Assumption why this happens

I only partially understand this. Since I selected the "PayPal" Expense Account, it seems like ff3 tries to find an Expense Account with the selected "PayPal" Expense Account-ID for all expenses (correct) AND a Revenue Account with the same "PayPal" Expense Account-ID for all revenues (not possible).
Since it's impossible to find a Revenue Account with the same ID, I assume it should search the Revenue Account with the name "PayPal" instead of the ID. This should be possible since account names of the same category have to be unique.
I think this should actually work like hinted in the error message but for some reason the name value is an empty string in this function call.
Is the name value this function needs the name that I'm expecting ("PayPal") or is this some other value?

---

Sorry for this lenghty explaination. I hope you get what I'm trying to explain here.
Can anyone reproduce this?
Will provide more debug information, data and screenshots if that's not enough.

Edit: for clarification: I'm trying to primarily track my bank account with ff3. PayPal is just the example I used. The same principle applies when you for example lend money to two different accounts of a friend and they pay you back from these accounts. If I understood ff3 correctly, if you use mapping, you would have one expense account and one revenue account with the name of your friend in this case.

**Comment by sodaws** [comment]
Yay, progress!
I got fidi to import everything without errors and do proper mapping.
Ok, so I did another fresh install, did the same steps but this time selected my "PayPal" Revenue Account instead of my "PayPal" Expense Account in fidi's mapping menu. For some reason it works this way but not the other way around.
Gonna gather some logs and create a more detailed guide for reproducing this with testdata...

**Reply by JC5** [reply]
I'm curious to see those. If a mapping is to an opposing account Firefly III should auto-correct for it.

**Reply by sodaws** [reply]
[ff3_debug.zip](https://github.com/firefly-iii/firefly-iii/files/9119931/ff3_debug.zip)

Ok, did more testing and collected everything.

### Logs
The zip file contains multiple logs. Each is of a fresh install.
* `relevant_lines.txt` for each log I noted the relevant lines for the transaction that's causing the errors, so that you don't have to.
* `revenue_success_log.txt` is the log of my answer, where I selected the Revenue Account in the fidi mapping menu.  Everything works as expected
* `expense_error_log.txt` is the log of the situation in my original post, where I selected the Expense Account in the fidi mapping menu.
* `expense_then_revenue_log.txt` is the log, if you first select the Expense Account first and then do a second import with the Revenue Account selected. If you do this, all 4 transactions are imported but the 2 from the first import are not recognised as duplicates and are also imported a second time. Resulting in 6 transaction in ff3.

### To recreate 
files mentioned here are in the zip
1. Do a fresh install
2. Create an account for logging in
3. Create an Asset Account named "ExampleAssetAccount"
4. Create a Expense Account named "PayPal" (other fields empty)
5. Create a Revenue Account named "PayPal" (other fields empty)
6. Create a personal access token, write it to fidi's env file and restart fidi
7. Open fidi -> import file
8. Select `testdata.csv` for importable file and `testconfig_expense_error.json` for optional configuration file -> hit next
9. Nothing to change on the configuration screen -> hit submit
10. Nothing to change on the role definition screen -> hit submit
11.  Now on the map screen, the mappings should be preselected but I always select a different one and then select the correct one again because I can't know if the selection that the configuration file did for me is the correct one. In case you created the Revenue Account before the Expense Account the mapping is false. Make sure you select the "PayPal" under the "Expense Account" section. -> hit submit
12. Start job -> Start job
13.  Now you should see this log:

    Error(s)
    
        Line #2:
            (2)transactions.0.source_id: Could not find a valid source account when searching for ID "4" or name "". (original value: "4")
            (2)transactions.0.source_name: Could not find a valid source account when searching for ID "4" or name "". (original value: "(not found)")
        Line #3:
            (2)transactions.0.source_id: Could not find a valid source account when searching for ID "4" or name "". (original value: "4")
            (2)transactions.0.source_name: Could not find a valid source account when searching for ID "4" or name "". (original value: "(not found)")
    
    Message(s)
    
        Line #0: Created withdrawal [#1 "PayPal Purchase #42"](http://firefly-iii-app-1:8080/transactions/show/1) (EUR 50)
        Line #1: Created withdrawal [#2 "PayPal Purchase #43"](http://firefly-iii-app-1:8080/transactions/show/2) (EUR 100)

To do the successful import, i do a fresh install and select `testconfig_revenue_success.json` this time.

**Reply by JC5** [reply]
Thanks. I found the issue. Because of the mapping, I have to fall back to submitting the original values if the opposing account type doesn't match. That means you may get a deposit from "PAYPAL EUROPE SARL ET CIE SCA", but at least it's stored.

I hope to release a new version soon with this in it.

**Reply by sodaws** [reply]
Awesome, thank you!

**Reply by JC5** [reply]
:+1: In the mean time, 0.9.12 came out which should include this fix!

**Reply by sodaws** [reply]
**EDIT: WAIT**, maybe I misunderstood your last message. Did you mean that you quick-fixed the error in 0.9.12 and want to restore both sides mapping in a later release anyway or did you mean that how it works now in 0.9.12 is the behavior that you want to leave it at?

EDIT2: Ah, I just saw that the Docker image for ff3-core was also updated...was still on the previsous version. Maybe have to test again tomorrow.
EDIT3: Nevermind, it was just the develop tag

Hey, I just tested the new version and can confirm that the error is fixed.

However, it seems like some functionality was unfortunately lost with 0.9.12 😅
With the older version (0.9.10 docker image) I was able to do mapping on both sides (Expense/Revenue) at the same time by selecting the Revenue Account during mapping. Meaning that the result was a single PayPal Revenue Account and a single PayPal Expense Account.

With the new version, mapping can only happen on one side during the import. Fixing the error of not being able to import the transaction was actually secondary to me, since I was able to do it how I wanted by selecting the Revenue Account anyway. I hoped that the new implementation would make mapping both sides at the same time during a single import work regardless of whether you selected the Revenue or Expense Account.

For now, I will continue to use the older version because both sides mapping works there. Fixing the error was good, don't get me wrong, but it would be super nice, if you could re-add the described behavior of both sides mapping in a later version.

**Reply by JC5** [reply]
I have no idea what you mean but I'll check it out 👍

**Reply by JC5** [reply]
I pushed some changes to develop. This is what I see happening:

* Withdrawals:
  * Destination account is unmapped: new expense account created by the same name
  * Destination account is mapped to an expense account: works as expected.
  * Destination account is mapped to an asset account: transaction transforms into a transfer
  * Destination account is mapped to a revenue account: a new expense account is made with the original value from the CSV as the name of the expense account.
* Deposits
  * Source account is unmapped: new revenue account created by the same name
  * Source account is mapped to an expense account: a new revenue account is made with the original value from the CSV as the name of the revenue account
  * Source account is mapped to an asset account: transaction transforms into a transfer
  * Source account is mapped to a revenue account: works as expected.

**Comment by sodaws** [comment, marked as answer]
Hey, I'm back, sorry to have kept you waiting...was busy with other stuff.
What do you think about adding a step in these two cases?

> Withdrawals:
 > * Destination account is mapped to a revenue account: a new expense account is made with the original value from the CSV as the name of the expense account.

> Deposits:
 > * Source account is mapped to an expense account: a new revenue account is made with the original value from the CSV as the name of the revenue account

Maybe something like this:

Withdrawals:
Destination account is mapped to a revenue account: 
1. Take the name of the revenue account and check if there exists a expense account with that name.
1.1. if yes, use that account
1.2. if no, create a new expense account with the name of the revenue account
2. If the above is for some reason not possible, fall back to the original procedure: a new expense account is made with the original value from the CSV as the name of the expense account.

And basically the same for Deposits:
Source account is mapped to an expense account:
1. Take the name of the expense account and check if there exists a revenue account with that name.
1.1. if yes, use that account
1.2. if no, create a new revenue account with the name of the expense account
2. If the above is for some reason not possible, fall back to the original procedure: a new revenue account is made with the original value from the CSV as the name of the revenue account

**Reply by JC5** [reply]
Thanks for the suggestion. This is partly in place and for the most part the data importer will correctly fall back to the opposing account type. Your described situation is a rare exception, because the selection dropdown already distinguishes between expense and revenue accounts.

Due to data quality of most banks, I try to make as few assumptions about the import as possible. This prevents magical behavior like you describe, where the data importer decides for you which account it has to be. So, whenever there's any doubt about what the right answer should be, the data importer will submit the content "as is".

The reason this happens is because 9 times out of 10, when I change something for the better for user A, it will break something else for user B. Any data choice I make will also influence other users, and more often than not for the worse.

So there will not be a "check if the account exists and when it does or maybe it doesn't than do something" check. If the mapping is wrong, and there is no obvious fix in the scope of the original user data, the original account name will be submitted.

**Reply by sodaws** [reply]
Funny, I imagined this to be a very common case, I guess it's actually not at all.
Yeah, I totally get the thing about making few assumptions, magical behavior and this circling around with changes that always break some users' workflows. 😅

While reading, I just realised the obvious solution to my scenario so that I don't have to manually do the mapping for these transactions.
I just have to split my CSV into two with something like pandas, creating one for withdrawals and one for deposits. Then import one after the other. I just map the withdrawals first and then the deposits or the other way around.
This way, I can do my mapping just how I wanted it without fidi having to do corrections and fallbacks.
